File: module/rank.js
/**
 * @fileoverview 酷我音乐排行榜内容接口
 *
 * 接口来源：default.zip 中 html/webdata/netsong/js/channel_bang.js 与 content_bang.js
 * 端点：https://kbangserver.kuwo.cn/ksong.s
 *
 * URL 模板（压缩包确认）：
 *   https://kbangserver.kuwo.cn/ksong.s?from=pc&fmt=json&type=bang&data=content
 *     &id={bangid}&pn={pn}&rn={rn}&isbang=1&show_copyright_off=0&pcmp4=1&bangid={bangid}
 *     &t={timestamp}
 *
 * 参数说明：
 * | 参数 | 值/格式 | 说明 |
 * | from | pc | 固定值 |
 * | fmt | json | 返回格式 |
 * | type | bang | 固定值 |
 * | data | content | 取榜单内容 |
 * | id | %d | 榜单 ID（16=热歌榜, 17=新歌榜, 93=飙升榜, 132=亚洲榜） |
 * | pn | %d | 页码，从 0 开始 |
 * | rn | %d | 每页数量（频道页 20，内容页 200） |
 * | isbang | 1 | 固定值 |
 * | show_copyright_off | 0 | 固定值 |
 * | pcmp4 | 1 | 固定值 |
 * | bangid | %d | 子榜单 ID（可选） |
 * | t | %d | 时间戳，防缓存 |
 *
 * 已知榜单 ID（来源：channel_bang.html c-id 属性）：
 * | id | sourceid | 名称 |
 * | 16 | 26 | 酷我热歌榜 |
 * | 17 | 25 | 酷我新歌榜 |
 * | 93 | 80358 | 酷我飙升榜 |
 * | 132 | -99 | 酷音乐亚洲排行榜 |
 *
 * 路由: /rank
 *
 * @module rank
 * @see REVERSE_SPEC.md 第三章第3.11节
 */

module.exports = (params, useAxios) => {
  const id = params?.id || params?.bangid || 16;

  const dataMap = {
    from: 'pc',
    fmt: 'json',
    type: 'bang',
    data: 'content',
    id,
    pn: params?.pn || 0,
    rn: params?.rn || 200,
    isbang: 1,
    show_copyright_off: 0,
    pcmp4: 1,
    t: Date.now(),
  };

  // bangid 子榜单（来源：content_bang.js 第 53、146 行）
  if (params?.bangid) {
    dataMap.bangid = params.bangid;
  }

  return useAxios({
    baseURL: 'https://kbangserver.kuwo.cn',
    url: '/ksong.s',
    method: 'GET',
    params: dataMap,
    cookie: params?.cookie || {},
    injectVer: false,
  });
};

File: module/rank.go
package module

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
	"kuwoapi/util"
)

// Rank 排行榜内容接口
// 来源：channel_bang.js + content_bang.js
// 端点：https://kbangserver.kuwo.cn/ksong.s
func Rank(params map[string]interface{}, r *http.Request) (map[string]interface{}, error) {
	id := util.GetString(params, "id", "16")
	pn := util.GetInt(params, "pn", 0)
	rn := util.GetInt(params, "rn", 200)

	u := fmt.Sprintf("https://kbangserver.kuwo.cn/ksong.s?from=pc&fmt=json&type=bang&data=content&id=%s&pn=%d&rn=%d&isbang=1&show_copyright_off=0&pcmp4=1&t=%d",
		id, pn, rn, util.GetInt(params, "t", 0))

	resp, err := util.HTTPClient.Get(u)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	body, _ := io.ReadAll(resp.Body)
	var result map[string]interface{}
	if err := json.Unmarshal(body, &result); err != nil {
		return map[string]interface{}{"code": 200, "data": string(body)}, nil
	}

	// 提取每首歌的 P2P 签名（param 字段，来源 channel_bang.js saveMusicInfo）
	if ml, ok := result["musiclist"].([]interface{}); ok {
		for _, c := range ml {
			if m, ok := c.(map[string]interface{}); ok {
				chain, _ := m["param"].(string)
				mergeParam(m, chain)
			}
		}
	}

	return map[string]interface{}{"code": 200, "data": result}, nil
}

// RankList 榜单列表接口
// 来源：channel_bang.js（排行榜频道页左侧分类树）
// 端点：http://qukudata.kuwo.cn/q.k?op=query&cont=tree&node=2&pn=0&rn=20&fmt=json&src=mbox&level=2
// 客户端逻辑：pc_extend 含 NOTSHOWPC2015 的项在 PC 隐藏；
//   pc_extend 中 BDTYPE-{分组}-{oid} 决定分组展示，缺省为「特色榜」；
//   sourceid 17/16/93 为特殊介绍项（客户端仅取 intro 文本）
func RankList(params map[string]interface{}, r *http.Request) (map[string]interface{}, error) {
	u := "http://qukudata.kuwo.cn/q.k?op=query&cont=tree&node=2&pn=0&rn=20&fmt=json&src=mbox&level=2"

	resp, err := util.HTTPClient.Get(u)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	body, _ := io.ReadAll(resp.Body)
	var result map[string]interface{}
	if err := json.Unmarshal(body, &result); err != nil {
		return map[string]interface{}{"code": 200, "data": string(body)}, nil
	}

	children, _ := result["child"].([]interface{})
	groups := make(map[string][]map[string]interface{})
	order := []string{}
	list := make([]map[string]interface{}, 0, len(children))
	for _, c := range children {
		item, ok := c.(map[string]interface{})
		if !ok {
			continue
		}
		ext, _ := item["pc_extend"].(string)
		if strings.Contains(ext, "NOTSHOWPC2015") {
			continue
		}
		sourceid := fmt.Sprintf("%v", item["sourceid"])

		group := ""
		hot := false
		for _, kv := range strings.Split(ext, "|") {
			switch {
			case kv == "HOT":
				hot = true
			case strings.HasPrefix(kv, "BDTYPE"):
				parts := strings.SplitN(kv, "-", 3)
				if len(parts) >= 2 {
					group = parts[1]
				}
			}
		}
		if group == "" && sourceid != "17" && sourceid != "16" && sourceid != "93" {
			group = "特色榜"
		}

		entry := map[string]interface{}{
			"id":        sourceid,
			"name":      item["name"],
			"pic":       item["pic"],
			"info":      item["info"],
			"intro":     item["intro"],
			"listen":    item["listen"],
			"isnew":     item["isnew"],
			"newcnt":    item["newcnt"],
			"hot":       hot,
			"special":   sourceid == "17" || sourceid == "16" || sourceid == "93",
		}
		list = append(list, entry)

		if group != "" {
			if _, seen := groups[group]; !seen {
				order = append(order, group)
			}
			groups[group] = append(groups[group], entry)
		}
	}

	groupedList := make([]map[string]interface{}, 0, len(order))
	for _, g := range order {
		groupedList = append(groupedList, map[string]interface{}{
			"group": g,
			"list":  groups[g],
		})
	}

	return map[string]interface{}{
		"code": 200,
		"data": map[string]interface{}{
			"name":   result["name"],
			"count":  len(list),
			"list":   list,
			"groups": groupedList,
		},
	}, nil
}
